README.md
# sortbuf -- data structure for sorting large numbers of items in memory

This library provides types and traits for accumulating a large number of items
in memory and iterating over them in ascending or descending order. It
outperforms `BTree`-based sorting, introduces low memory overhead and allows
insertion of items from multiple threads as well as reacting to allocation
failures without losing data. However, it's sole purpose is sorting and it
provides no other functionality.

## Example

```rust
let mut sortbuf = sortbuf::SortBuf::new();
let mut inserter = sortbuf::Inserter::new(&mut sortbuf);
inserter.insert_items([10, 20, 5, 17]).expect("Failed to insert items");
drop(inserter);
assert!(sortbuf.into_iter().eq([20, 17, 10, 5]));
```

## License

This work is provided under the MIT license. See `LICENSE` for more details.



CHANGELOG.md
# 0.1.0 -- 2022-06-26

## Added
- Initial version of the library, including documentation, tests as well as a
  `sort` example and a `comp` benchmark.


src/tests.rs
// SPDX-License-Identifier: MIT
/// Tests

use super::*;

use std::cmp::Reverse;
use std::num::NonZeroUsize;

use rand::Rng;


/// Item type to use for testing
type Item = u64;


#[test]
fn inserter_simple() {
    let mut buf: SortBuf<_> = Default::default();
    {
        let mut inserter = inserter::Inserter::new(&mut buf);
        inserter.set_bucket_size(NonZeroUsize::new(1000).expect("Failed to construct bucket size"));
        inserter.extend(random_items(10_500));
    }

    let iter = buf.into_iter();
    assert_eq!(iter.len(), 10_500);
}


#[test]
fn inserter_half() {
    let mut buf: SortBuf<_> = Default::default();
    {
        let mut inserter = inserter::Inserter::new(&mut buf);
        inserter.set_bucket_size(NonZeroUsize::new(1000).expect("Failed to construct bucket size"));
        inserter.extend(random_items(500));
    }

    let iter = buf.into_iter();
    assert_eq!(iter.len(), 500);
}


#[test]
fn inserter_exact() {
    let mut buf: SortBuf<_> = Default::default();
    {
        let mut inserter = inserter::Inserter::new(&mut buf);
        inserter.set_bucket_size(NonZeroUsize::new(1000).expect("Failed to construct bucket size"));
        inserter.extend(random_items(10_000));
    }

    let iter = buf.into_iter();
    assert_eq!(iter.len(), 10_000);
}


#[test]
fn inserter_multiple() {
    let mut buf: SortBuf<_> = Default::default();
    {
        let mut inserter = inserter::Inserter::new(&mut buf);
        inserter.set_bucket_size(NonZeroUsize::new(1000).expect("Failed to construct bucket size"));
        inserter.extend(random_items(500));
        inserter.extend(random_items(1000));
    }

    let iter = buf.into_iter();
    assert_eq!(iter.len(), 1500);
}


#[test]
fn iter_sorted() {
    let mut items = random_items(10_500);

    let iter: iter::Iter<Item> = std::iter::from_fn(
        move || Some(items.by_ref().take(1000).collect::<Vec<_>>())
    ).take_while(|v| !v.is_empty()).map(bucket::Bucket::new).map(Into::into).collect::<Vec<_>>().into();

    assert_sorted(iter.map(Reverse))
}

#[test]
fn iter_count() {
    let mut items = random_items(10_500);

    let iter: iter::Iter<Item> = std::iter::from_fn(
        move || Some(items.by_ref().take(1000).collect::<Vec<_>>())
    ).take_while(|v| !v.is_empty()).map(bucket::Bucket::new).map(Into::into).collect::<Vec<_>>().into();

    assert_eq!(iter.count(), 10_500)
}


#[test]
fn bucket_sorted() {
    let bucket = bucket::Bucket::new(random_items(1000).collect::<Vec<_>>());
    assert_sorted(bucket::SortedBucket::from(bucket).map(Reverse))
}


/// Construct an [Iterator] yielding `num` random items
fn random_items(num: usize) -> impl Iterator<Item = Item> {
    let mut rng = rand_pcg::Mcg128Xsl64::new(0xcafef00dd15ea5e5); // seed taken from rand_pcg docs
    std::iter::from_fn(move || Some(rng.gen::<Item>())).take(num)
}

/// Check whether the given [Iterator] is sorted
fn assert_sorted<T: Ord>(mut iter: impl Iterator<Item = T>) {
    if let Some(first) = iter.next() {
        iter.try_fold(first, |c, n| if c <= n { Some(n) } else { None })
            .expect("Iterator does not yield sorted items");
    }
}



src/lib.rs
// SPDX-License-Identifier: MIT
//! Sort a large number of items in memory
//!
//! This library provides types, most prominently [SortBuf], and traits for
//! accumulating a large number of items in memory and iterating over them in
//! ascending or descending order (as per [Ord]). The implementation
//!
//! * avoids potentially costly reallocations,
//! * releases chunks of memory every now and then during iteration,
//! * doesn't introduce much memory overhead,
//! * allows reacting to allocation failures,
//! * supports multi-threaded insertion and
//! * isn't awfully slow.
//!
//! # Examples
//!
//! Natively, [SortBuf] will prepare items for descending iteration:
//!
//! ```
//! let mut sortbuf = sortbuf::SortBuf::new();
//! let mut inserter = sortbuf::Inserter::new(&mut sortbuf);
//! inserter.insert_items([10, 20, 5, 17]).expect("Failed to insert items");
//! drop(inserter);
//! assert!(sortbuf.into_iter().eq([20, 17, 10, 5]));
//! ```
//!
//! For ascending iteration, items need to be wrapped in [std::cmp::Reverse].
//! However, the library provides convenience functions for handling the
//! (un)wrapping:
//!
//! ```
//! let mut sortbuf = sortbuf::SortBuf::new();
//! let mut inserter = sortbuf::Inserter::new(&mut sortbuf);
//! inserter.insert_items_reversed([10, 20, 5, 17]).expect("Failed to insert items");
//! drop(inserter);
//! assert!(sortbuf.unreversed().eq([5, 10, 17, 20]));
//! ```
//!
//! Multithreaded insertion is supported via multiple [Inserter]s:
//!
//! ```
//! use std::sync::{Arc, Mutex};
//! let sortbuf: Arc<Mutex<sortbuf::SortBuf<_>>> = Default::default();
//! let workers: Vec<_> = (0..4).map(|n| {
//!     let mut inserter = sortbuf::Inserter::new(sortbuf.clone());
//!     std::thread::spawn(move || inserter
//!         .insert_items((0..1000).map(|i| 4*i+n))
//!         .expect("Failed to insert items"))
//! }).collect();
//! workers.into_iter().try_for_each(|h| h.join()).unwrap();
//! assert!(sortbuf.lock().unwrap().take().into_iter().eq((0..4000).rev()));
//! ```
//!
//! # Approach and comparison
//!
//! As indicated in the examples above, adding new items to a buffer is done via
//! [Inserter]s. These accumulate items in pre-sorted [Bucket]s and commit them
//! to their target buffer. Later, that buffer can be converted to an [Iterator]
//! which yields items taken from those [Bucket]s, which involves selecting the
//! [Bucket] with the current greatest item in the buffer.
//!
//! While a significant amount of time is spent during insertion, the majority
//! of time is usually spent during iteration. Performance is usually better,
//! and skews towards more time spent in the parallelizable insertion state,
//! with fewer, bigger [Bucket]s. As [Bucket]s are pre-allocated, this comes at
//! the cost of flexibility regarding memory.
//!
//! ## Comparison to Vec and sort
//!
//! Buffering and sorting items can also be done using a [Vec] (for buffering)
//! in conjunction with [slice::sort], [slice::sort_unstable] or another sorting
//! function. The process is then usually split into an insertion, a sorting and
//! an iteration phase, with sort being the most computational intensive phase.
//!
//! Sorting and iteration over the items in a [Vec] is generally faster than
//! with the utilities provided by this library ---in the single-threaded case.
//! However, this library does allow insertion from multiple threads, contrary
//! to a bare [Vec]. In addition, the performance of inserting items into a
//! [Vec] hinges on the reallocation performance, which might be poor in some
//! cases, e.g. if multiple buffers are involved.
//!
//! The need of a single, separate, computational intensive sorting phase may
//! also have some implications on overall performance in some use-cases. With
//! the types provided by this library, sorting will likely interleave with I/O
//! linked to insertion and/or the final iteration, spread out over the entire
//! process. Thus, the underlying OS may have more opportunities to perform
//! background operations related to reads (insertion stage) and writes
//! (iteration stage), increasing the overall throughput.
//!
//! ## Comparison to BTreeSet
//!
//! Another option for sorting items without the need for a separate sorting
//! phase would be an [BTreeSet](std::collections::BTreeSet). Contrary to the
//! `sortbuf` approach, most of the time is spent in the insertion phase rather
//! than the iteration phase. Using a [BTreeSet](std::collections::BTreeSet) is
//! usually slower than a [SortBuf] with sufficiently large [Bucket]s, not
//! parallelizable and incurs a higher memory overhead.

#![feature(allocator_api)]

use std::alloc::{Allocator, Global};

mod bucket;
mod inserter;
mod iter;

pub mod error;

#[cfg(test)]
mod tests;


pub use bucket::{Bucket, DEFAULT_BUCKET_BYTESIZE};
pub use inserter::{BucketAccumulator, Inserter};


/// Data structure for preparing a large number of items for sorted iteration
///
/// This data structure buffers items for later iteration in descending order.
/// New items are inserted via an [Inserter] which has to be constructed
/// separately for a given buffer. Once all (or sufficently many) items are
/// inserted, [IntoIterator] may be used for iterating over these items in
/// descending order (according to the items' implementation of [Ord]).
///
/// For ascending iteration, users need to wrap items in [std::cmp::Reverse] and
/// unwrap them during the final iteration.
///
/// # Time complexity
///
/// Assuming a fixed bucket size _b_, the estimated runtime cost of inserting _n_
/// items via [Inserter]s is O(_n_ log(_b_)). The estimated runtime cost of
/// draining the [Iterator] provided through this type's [IntoIterator] impl is
/// O(_n_ log(_n_/_b_)).
///
/// In practice, the runtime cost will be higher for draining the iterator than
/// for the insertion. The iteration performance will be significantly affected
/// by the term log(_n_/_b_), since it will be reflected in the average number
/// of cache-misses involved in retrieving a single item. Thus, greater values
/// of _b_ are to be preferred. As a rule of thumb, aim for a number of buckets
/// (i.e. _n_/_b_) well under 100 for ok-ish performance.
///
/// Note also that the cost of insertion can be split between multiple threads,
/// if the [SortBuf] is wrapped in a mutex for which the [BucketAccumulator]
/// trait is implemented.
///
/// # Other notes
///
/// The omission of an implementation of [Clone] for this type is on purpose, as
/// it is meant for large amounts of data.
#[derive(Debug)]
pub struct SortBuf<T: Ord, A: Allocator = Global> {
    buckets: Vec<bucket::SortedBucket<T, A>>,
}

impl<T: Ord> SortBuf<T, Global> {
    /// Create a new sorting buffer
    pub fn new() -> Self {
        Default::default()
    }
}

impl<T: Ord, A: Allocator> SortBuf<T, A> {
    /// Take this buffer's contents, leaving an empty buffer
    pub fn take(&mut self) -> Self {
        std::mem::take(self)
    }
}

impl<T: Ord, A: Allocator> Default for SortBuf<T, A> {
    fn default() -> Self {
        Self {buckets: Vec::new()}
    }
}

impl<T: Ord, A: Allocator> SortBuf<std::cmp::Reverse<T>, A> {
    /// Convert into an [Iterator] over items unwrapped from [std::cmp::Reverse]
    ///
    /// This funtion allows convenient retrieval of the buffered items in their
    /// unreversed order. Use this function if you need an iterator over items
    /// in ascending order.
    pub fn unreversed(self) -> impl Iterator<Item = T> {
        self.into_iter().map(|std::cmp::Reverse(v)| v)
    }
}

impl<T: Ord, A: Allocator> IntoIterator for SortBuf<T, A> {
    type Item = T;
    type IntoIter = iter::Iter<Self::Item, A>;

    fn into_iter(self) -> Self::IntoIter {
        self.buckets.into()
    }
}



src/iter.rs
// SPDX-License-Identifier: MIT
//! [Iter] type and related utilities

use std::alloc::{Allocator, Global};
use std::collections::binary_heap::{self, BinaryHeap};
use std::iter::FusedIterator;

use super::bucket::SortedBucket;


/// Default shrinking theshold in bytes
const DEFAULT_SHRINK_THRESHOLD_BYTES: usize = 1024*1024;


/// [Iterator] yielding items in descending order
///
/// This [Iterator] will yield an item only after all items greater have been
/// yielded.
///
/// The iterator will release memory from time to time during iteration. The
/// specifics are controlled via an internal threshold which can be altered
/// through [Iter::with_shrink_threshold] and
/// [Iter::with_shrink_threshold_bytes].
///
/// # Time complexity
///
/// The implementation of [Iterator::next] has an amortized time complexity of
/// O(log(_n_/_b_)) with _n_ denoting the number of items and _b_ denoting the
/// bucket size the instance was constructed with, under the assuption that the
/// distribution of values amongst buckets is uniform. Draining the entire
/// [Iterator] thus has an expected runtime complexity of O(_n_ log(_n_/_b_)).
///
/// # Other notes
///
/// The omission of an implementation of [Clone] for this type is on purpose, as
/// it is meant for large amounts of data.
#[derive(Debug)]
pub struct Iter<T: Ord, A: Allocator = Global> {
    buckets: BinaryHeap<SortedBucket<T, A>>,
    shrink_theshold: usize,
}

impl<T: Ord, A: Allocator> Iter<T, A> {
    /// Set the number of unused item slots buckets are allowed to accumulate
    ///
    /// This iterator pulls items from a number of buckets, which will thus
    /// accumulate unused item slots. If a certain number of unused slots exists
    /// in a bucket, the iterator will try to shrink the underlying storage and
    /// thus make memory availible again.
    ///
    /// This function allows specifying the shrinking threshold.
    pub fn with_shrink_threshold(self, shrink_theshold: usize) -> Self {
        Self{shrink_theshold, ..self}
    }

    /// Set the number of unused bytes buckets are allowed to accumulate
    ///
    /// This iterator pulls items from a number of buckets, which will thus
    /// accumulate unused memory. If a certain amount exists in a bucket, the
    /// iterator will try to shrink the underlying storage and thus make memory
    /// availible again.
    ///
    /// This function allows specifying the shrinking threshold, in bytes.
    pub fn with_shrink_threshold_bytes(self, shrink_theshold: usize) -> Self {
        self.with_shrink_threshold(shrink_theshold / std::mem::size_of::<T>())
    }
}

impl<T: Ord, A: Allocator> From<Vec<SortedBucket<T, A>>> for Iter<T, A> {
    fn from(buckets: Vec<SortedBucket<T, A>>) -> Self {
        Self{
            buckets: buckets.into(),
            shrink_theshold: DEFAULT_SHRINK_THRESHOLD_BYTES / std::mem::size_of::<T>(),
        }
    }
}

impl<T: Ord, A: Allocator> ExactSizeIterator for Iter<T, A> {}

impl<T: Ord, A: Allocator> FusedIterator for Iter<T, A> {}

impl<T: Ord, A: Allocator> Iterator for Iter<T, A> {
    type Item = T;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some(mut bucket) = self.buckets.peek_mut() {
            if let Some(item) = bucket.next() {
                if bucket.overcapacity() >= self.shrink_theshold {
                    bucket.shink_to_fit()
                }
                return Some(item)
            } else {
                binary_heap::PeekMut::pop(bucket);
            }
        }

        None
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let size = self.buckets.iter().map(ExactSizeIterator::len).sum();
        (size, Some(size))
    }
}



src/error.rs
// SPDX-License-Identifier: MIT
//! Types and utilities related to error handling and reporting

use std::collections::TryReserveError;
use std::error::Error;
use std::fmt;


/// Insertion error
///
/// This type conveys errors occuring during the insertion of items to a buffer.
#[derive(Debug)]
pub struct InsertionError(TryReserveError);

impl From<TryReserveError> for InsertionError {
    fn from(inner: TryReserveError) -> Self {
        Self(inner)
    }
}

impl Error for InsertionError {
    fn cause(&self) -> Option<&(dyn Error + 'static)> {
        Some(&self.0)
    }
}

impl fmt::Display for InsertionError {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.write_str("Could not add items to accumulator")
    }
}


/// Result type for insertions
///
/// In the case of an insertion failure, we need to return an `InsertionError`.
/// However, we also want to return the item(s) we failed to insert in some way.
pub type InsertionResult<T> = Result<(), (InsertionError, T)>;



src/bucket.rs
// SPDX-License-Identifier: MIT
//! Types representing individual buckets and related utilities

use std::alloc::{Allocator, Global};
use std::cmp::Ordering;
use std::fmt;


/// Default size for [Bucket]s
///
/// This constant holds a default size for buckets, in bytes. The constant
/// is choosen to be reasonably large without obstructing the library's use
/// on smaller machines. Currently, it is set to 16MB.
///
/// The rationale behind that value is that on a typical SBC with a quadcore
/// and 1GB of ram, it should be possible to accumulate items into buckets
/// for multiple (e.g. 3) buffers on all cores without exhausting memory (or
/// running into overcommitting).
pub const DEFAULT_BUCKET_BYTESIZE: usize = 16*1024*1024;


/// A collection of items to be committed to a [SortBuf](super::SortBuf)
///
/// Users of the library will usually not use this type directly.
///
/// # Other notes
///
/// The omission of an implementation of [Clone] for this type is on purpose, as
/// it holds non-shared ownership over significant amounts of data.
pub struct Bucket<T, A: Allocator = Global>(Vec<T, A>);

impl<T: Ord, A: Allocator> Bucket<T, A> {
    /// Create a bucket from a [Vec] of items
    ///
    /// # Time complexity
    ///
    /// Construction of a sorted bucket involves sorting the items. Thus, it
    /// comes with a run-time cost of O(_b_*log(_b_)) with bucket size _b_.
    pub(crate) fn new(mut items: Vec<T, A>) -> Self {
        items.shrink_to_fit();
        items.sort_unstable();
        Self(items)
    }

    /// Convert this bucket back to a [Vec]
    pub(crate) fn into_inner(self) -> Vec<T, A> {
        self.0
    }

    /// Retrieve the number of items in this bucket
    pub fn len(&self) -> usize {
        self.0.len()
    }
}

impl<T: Ord, A: Allocator> fmt::Debug for Bucket<T, A> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(fmt, "Bucket({} items)", self.len())
    }
}


/// A sorted collection of items
///
/// This type wraps a [Vec] of items sorted in ascending order and implements
/// [Ord] based on its last element. The ordering amongst buckets of this type
/// is equivalent to the ordering of the maximum item in each bucket.
///
/// In addition, a `SortedBucket` functions as an [Iterator] yielding (and
/// removing) its elements from last to first, i.e. in reverse or descending
/// order.
///
/// # Other notes
///
/// The omission of an implementation of [Clone] for this type is on purpose, as
/// it holds non-shared ownership over significant amounts of data.
pub(crate) struct SortedBucket<T: Ord, A: Allocator>(Vec<T, A>);

impl<T: Ord, A: Allocator> SortedBucket<T, A> {
    /// Retrieve the current overcapacity of this bucket
    ///
    /// The overcapacity is defined as the number of additional items the inner
    /// [Vec] has capacity for.
    #[inline(always)]
    pub fn overcapacity(&self) -> usize {
        self.0.capacity() - self.0.len()
    }

    /// Shrink the inner [Vec] to the number of items it currently holds
    ///
    /// This operation sheds overcapacity.
    #[inline(always)]
    pub fn shink_to_fit(&mut self) {
        self.0.shrink_to_fit()
    }
}

impl<T: Ord, A: Allocator> From<Bucket<T, A>> for SortedBucket<T, A> {
    fn from(Bucket(items): Bucket<T, A>) -> Self {
        Self(items)
    }
}

impl<T: Ord, A: Allocator> ExactSizeIterator for SortedBucket<T, A> {}

impl<T: Ord, A: Allocator> std::iter::FusedIterator for SortedBucket<T, A> {}

impl<T: Ord, A: Allocator> Iterator for SortedBucket<T, A> {
    type Item = T;

    #[inline(always)]
    fn next(&mut self) -> Option<T> {
        self.0.pop()
    }

    #[inline(always)]
    fn size_hint(&self) -> (usize, Option<usize>) {
        let len = self.0.len();
        (len, Some(len))
    }
}

impl<T: Ord, A: Allocator> Ord for SortedBucket<T, A> {
    #[inline(always)]
    fn cmp(&self, other: &Self) -> Ordering {
        Ord::cmp(&self.0.last(), &other.0.last())
    }
}

impl<T: Ord, A: Allocator> PartialOrd for SortedBucket<T, A> {
    #[inline(always)]
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        PartialOrd::partial_cmp(&self.0.last(), &other.0.last())
    }
}

impl<T: Ord, A: Allocator> Eq for SortedBucket<T, A> {}

impl<T: Ord, A: Allocator> PartialEq for SortedBucket<T, A> {
    #[inline(always)]
    fn eq(&self, other: &Self) -> bool {
        PartialEq::eq(&self.0.last(), &other.0.last())
    }
}

impl<T: Ord, A: Allocator> fmt::Debug for SortedBucket<T, A> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(fmt, "SortedBucket({} items)", self.len())
    }
}



examples/sort.rs
// SPDX-License-Identifier: MIT
//! Simplistic impl of unix `sort` utility
//!
//! This program prints the line-wise sorted concatenation of its inputs. Inputs
//! are supplied as paths via command line arguments. If no inputs are supplied,
//! the program sorts its standart input.


fn main() {
    use std::io::BufRead;
    use std::sync::{Arc, Mutex};

    // We need to pre-collect the arguments (minus the progname) since
    // ArgsOs is both `!Send` and `!Sync`.
    let args: Vec<_> = std::env::args_os().skip(1).collect();

    let lines: Arc<Mutex<sortbuf::SortBuf<_>>> = Default::default();

    if args.len() > 0 {
        let paths: Arc<Mutex<_>> = Mutex::new(args.into_iter()).into();

        // We delegate the actual work to multiple worker threads
        let workers: Vec<_> = (0..std::thread::available_parallelism().map(|n| n.get()).unwrap_or(2))
            .map(|_| {
                let paths = paths.clone();
                let lines = lines.clone();
                std::thread::spawn(move || {
                    let mut inserter = sortbuf::Inserter::new(lines);
                    while let Some(path) = paths.lock().unwrap().next() {
                        let lines = std::io::BufReader::new(std::fs::File::open(path).unwrap())
                            .lines()
                            .map(|l| l.unwrap());
                        inserter.insert_items_reversed(lines).unwrap()
                    }
                })
            }).collect();

        // Wait for workers to finish...
        workers.into_iter().try_for_each(|t| t.join()).unwrap()
    } else {
        sortbuf::Inserter::new(lines.clone())
            .insert_items_reversed(std::io::stdin().lock().lines().map(|l| l.unwrap()))
            .unwrap()
    }

    lines.lock().unwrap().take().unreversed().for_each(|l| println!("{}", l));
}



benches/comp.rs
// SPDX-License-Identifier: MIT
//! Comparison between different kinds of sorting buffers

use std::sync::atomic;
use std::time::Duration;


const NUM_THREADS: usize = 4;
const MAX_M_ITEMS: usize = 1024;


fn main() {
    println!("implementation | 2^20 Is | T wall  | T_f usr | T_f sys | T_d usr | T_d sys | mem     ");
    println!("---------------|---------|---------|---------|---------|---------|---------|---------");


    let benches: [(_, &dyn Fn(usize) -> (Duration, Diff, Diff)); 6] = [
        ("baseline",        &|i| bench_func(baseline, i)),
        ("vec",             &|i| bench_func(fill_vec, i)),
        ("btree",           &|i| bench_func(fill_btree, i)),
        ("sortbuf",         &|i| bench_func(fill_sortbuf, i)),
        ("sortbuf jumbo",   &|i| bench_func(fill_sortbuf_jumbo, i)),
        ("sortbuf 4t",      &|i| bench_func(fill_sortbuf_threads, i)),
    ];

    std::iter::successors(Some(1usize), |s| (*s).checked_mul(4))
        .take_while(|s| *s <= MAX_M_ITEMS)
        .flat_map(|s| benches.iter().map(move |(n, b)| (n, s, b)))
        .for_each(|(n, s, b)| {
            let (t, d1, d2) = b(s*1024*1024);
            println!(
                "{:<15}|{:>9}|{:>9}|{:>9}|{:>9}|{:>9}|{:>9}|{:>9}",
                n,
                s,
                t.as_millis(),
                d1.user_time.as_millis(),
                d1.system_time.as_millis(),
                d2.user_time.as_millis(),
                d2.system_time.as_millis(),
                d1.allocated / (1024*1024),
            )
        });
}


fn baseline(num: usize) -> impl IntoIterator<Item=u64> {
    let mut curr = Default::default();
    random_items(num).map(move |v| {
        curr = v.saturating_add(curr);
        curr
    })
}


fn fill_vec(num: usize) -> impl IntoIterator<Item=u64> {
    let mut buf: Vec<_> = random_items(num).collect();
    buf.sort_unstable();
    buf
}


fn fill_btree(num: usize) -> impl IntoIterator<Item=u64> {
    random_items(num).collect::<std::collections::BTreeSet<_>>()
}


fn fill_sortbuf(num: usize) -> impl IntoIterator<Item=u64> {
    let mut buf: sortbuf::SortBuf<_> = Default::default();

    let mut inserter = sortbuf::Inserter::new(&mut buf);
    inserter.insert_items_reversed(random_items(num)).expect("Error while inserting");
    std::mem::drop(inserter);

    buf.unreversed()
}


fn fill_sortbuf_jumbo(num: usize) -> impl IntoIterator<Item=u64> {
    let mut buf: sortbuf::SortBuf<_> = Default::default();

    let mut inserter = sortbuf::Inserter::new(&mut buf);
    inserter.set_bucket_bytesize(sortbuf::DEFAULT_BUCKET_BYTESIZE * 4);
    inserter.insert_items_reversed(random_items(num)).expect("Error while inserting");
    std::mem::drop(inserter);

    buf.unreversed()
}


fn fill_sortbuf_threads(num: usize) -> impl IntoIterator<Item=u64> {
    use std::sync::{Arc, Mutex};

    let buf: Arc<Mutex<sortbuf::SortBuf<std::cmp::Reverse<u64>>>> = Default::default();

    random_items(NUM_THREADS).map(|seed| {
        let mut inserter = sortbuf::Inserter::new(buf.clone());
        std::thread::spawn(move || inserter.insert_items_reversed(
            random_items_with_seed(num / NUM_THREADS, seed.into())
        ).expect("Error while inserting"))
    }).collect::<Vec<_>>().into_iter().try_for_each(|h| h.join()).expect("Error while waiting for threads");

    Arc::try_unwrap(buf)
        .expect("Failed to unwrap buffer!")
        .into_inner()
        .expect("Failed to unlock buffer!")
        .unreversed()
}


/// Benchmark a single sorting buffer
///
/// This function accepts a function which is expected to construct a sorted
/// buffer with random data as well as the number of items with which to fill
/// the buffer.
///
/// The function will construct the sorted buffer, measuring the resources used
/// suring the process. It will then drain the buffer, again measuring the
/// resource usage. Both resource usages are then returned to the caller
/// together with the overall wallclock time.
fn bench_func<I>(mk_iter: impl Fn(usize) -> I, items: usize) -> (Duration, Diff, Diff)
where I: IntoIterator<Item = u64>,
{
    let t0 = std::time::Instant::now();
    let s0 = Snapshot::new();
    let iter = mk_iter(items);

    let s1 = Snapshot::new();
    let mut iter = iter.into_iter();
    if let Some(first) = iter.next() {
        // TODO: use a black box of some sort to prevent the draining to be optimized out
        iter.try_fold(first, |c, n| if c <= n { Some(n) } else { None })
            .expect("Iterator does not yield sorted items");
    }

    let s2 = Snapshot::new();

    (t0.elapsed(), s1.diff(s0), s2.diff(s1))
}


/// Create an iterator over random items
fn random_items(num: usize) -> impl Iterator<Item = u64> {
    random_items_with_seed(num, 0xcafef00dd15ea5e5) // seed taken from rand_pcg docs
}


/// Create an iterator over random items with a given seed
fn random_items_with_seed(num: usize, seed: u128) -> impl Iterator<Item = u64> {
    use rand::Rng;

    let mut rng = rand_pcg::Mcg128Xsl64::new(seed);
    std::iter::from_fn(move || Some(rng.gen())).take(num)
}


/// Snapshot of the current resource usage and time
#[derive(Copy, Clone, Debug)]
struct Snapshot {
    user_time: Duration,
    system_time: Duration,
    allocated: usize,
}

impl Snapshot {
    /// Create a new [Snapshot]
    pub fn new() -> Self {
        let mut rusage = libc::rusage{
            ru_utime:       libc::timeval{tv_sec: 0, tv_usec: 0},
            ru_stime:       libc::timeval{tv_sec: 0, tv_usec: 0},
            ru_maxrss:      0,
            ru_ixrss:       0,
            ru_idrss:       0,
            ru_isrss:       0,
            ru_minflt:      0,
            ru_majflt:      0,
            ru_nswap:       0,
            ru_inblock:     0,
            ru_oublock:     0,
            ru_msgsnd:      0,
            ru_msgrcv:      0,
            ru_nsignals:    0,
            ru_nvcsw:       0,
            ru_nivcsw:      0,
        };
        if unsafe { libc::getrusage(libc::RUSAGE_SELF, &mut rusage) } != 0 {
            panic!("Failed to retrieve resource usages.")
        }

        Self{
            user_time: duration_from_timeval(rusage.ru_utime),
            system_time: duration_from_timeval(rusage.ru_stime),
            allocated: ALLOCATOR.allocated(),
        }
    }

    /// Compare this [Snapshot] to an earlier one
    pub fn diff(&self, older: Self) -> Diff {
        Diff{
            user_time: self.user_time - older.user_time,
            system_time: self.system_time - older.system_time,
            allocated: self.allocated.saturating_sub(older.allocated)
        }
    }
}


/// The difference between two [Snapshot]s
#[derive(Copy, Clone, Debug)]
struct Diff {
    pub user_time: Duration,
    pub system_time: Duration,
    pub allocated: usize,
}


/// Convert a [libc::timeval] to an [std::time::Duration]
fn duration_from_timeval(val: libc::timeval) -> Duration {
    Duration::new(
        val.tv_sec.try_into().expect("Timeval has unsuitable seconds."),
        (val.tv_usec * 1000).try_into().expect("Timeval has unsuitable microseconds."),
    )
}


struct AccountingAlloc {
    inner: std::alloc::System,
    allocated: atomic::AtomicUsize,
}

impl AccountingAlloc {
    /// Retrieve the number of bytes currently allocated via this allocator.
    fn allocated(&self) -> usize {
        self.allocated.load(atomic::Ordering::SeqCst)
    }
}

unsafe impl std::alloc::GlobalAlloc for AccountingAlloc {
    unsafe fn alloc(&self, layout: std::alloc::Layout) -> *mut u8 {
        self.allocated.fetch_add(layout.size(), atomic::Ordering::Release);
        self.inner.alloc(layout)
    }

    unsafe fn dealloc(&self, ptr: *mut u8, layout: std::alloc::Layout) {
        self.allocated.fetch_sub(layout.size(), atomic::Ordering::Release);
        self.inner.dealloc(ptr, layout)
    }

    unsafe fn realloc(&self, ptr: *mut u8, layout: std::alloc::Layout, new_size: usize) -> *mut u8 {
        if let Some(diff) = layout.size().checked_sub(new_size) {
            self.allocated.fetch_sub(diff, atomic::Ordering::Release);
        } else {
            self.allocated.fetch_add(new_size.saturating_sub(layout.size()), atomic::Ordering::Release);
        }
        self.inner.realloc(ptr, layout, new_size)
    }
}


#[global_allocator]
static ALLOCATOR: AccountingAlloc = AccountingAlloc{inner: std::alloc::System, allocated: atomic::AtomicUsize::new(0)};

